### DISCUSSION TITLE
使用用户态TCP/UDP协议栈的好处是啥

### DISCUSSION THREAD

**Discussion by ccdd4ever** [xjasonlyu/tun2socks#124] [Q&A] [answered]
如题，请问下使用gVisor的用户态TCP协议栈的意图是啥？有啥好处，有啥弊端

**Comment by zhu327** [comment, marked as answer]
tun网卡返回的数据需要在用户态直接解析以获得不错的性能，如果走内核态协议栈，需要重新启动一个端口监听，不断的用户态陷入内核态，再中断回到用户态，期间内存不断的复制，影响性能，可以看下项目的bench测试，用户态的协议栈的优势在此，这是我的理解
